### ISSUE TITLE
focus first element, regardless of whether or not its in the viewport

### ISSUE THREAD

**Issue by gruns** [MikesGlitch/Focus-First-Input#4]
first, i love your extension. thank you @MikesGlitch! 🙏 

now -- the default behavior i want 90% of the time is not to focus the first input element in the viewport, but instead to focus the first input element in the page, regardless of whether its in the viewport or not. that is, 90% of the time i want to focus some top bar input element on the page

insofar, please

1) change the default behavior to A) focus the first input ele in the page, not just in the viewport, and B) scroll the browser to that now-focused input ele. that is, remove the `elementInViewport()` check and add a scroll action to move the now-focused input ele into the viewport.

or

2) add a second, different, customizable keybind that calls `focusOnFirstInput()` without the `elementInViewport()` check and scrolls the new-focused input ele into the viewport. if helpful, this is the only keybind i would use with this extension

an additional nice-to-have QoL feature would be to graphically blink the now-focused input element subtly to draw attention to it. eg by pulsing a colored border for 1s. i can add this in a separate issue or PR

thank you! 🙌

**Comment by MikesGlitch** [comment]
Hi @gruns, thanks for the feedback and contribution 🎉 

Another option might be to keep the shortcut, but if the first input in viewport is already focused pressing again could change focus to the first input on the page.

Do you think this would be satisfy it or do you prefer another shortcut? If so, which shortcut should be the default?

**Comment by gruns** [comment]
hey @mikesglitch!

in your usage, do you more often want to focus the first input ele in the viewport and not the first input ele on the page?

> Another option might be to keep the shortcut, but if the first input in viewport is already focused pressing again could change focus to the first input on the page.

clever thinking! i love it. 👏 that said, I'm not a huge fan of this because
- often times i (and im sure others) will accidentally tap the same chord twice (or more) inadvertantly, and
- 99% of the time i want to focus the first input element on the page, not in the viewport. so this keybind setup would mean i have to tap the ctrl+i twice every invocation instead of only once

> Do you think this would be satisfy it or do you prefer another shortcut? If so, which shortcut should be the default?

for my own personal use, i would only bind focusing the first input element on the page. i rarely, if ever?, want to focus the non-first input element on the page

so the *best* approach would be to support it all:
1) support two separate keybinds:
  - one keybind to focus the first input ele in the viewport, and
  - another keybind to focus the first input ele on the page (what i want 99% of the time)
2) have an option for your clever idea above: if ctrl+i (or whatever the user rebinds it to) is tapped twice in 400ms, it performs the opposite selection

but that's more work. and i only need/want #1 above. so i say we just ship #1. and if someone wants #2 later? that can be added then. only build what people actually use, keep things small, and ship 🙂

but! as for the bind to jump to first input ele in the whole page? that's tricky because there are already so many ctrl+? binds taken across the browser + OS

so, again, what's the literal easiest/fastest thing to ship?
- keep ctrl+i as the bind. i like it and am already used to it 👍 
- add a simple toggle in the cfg that changes ctrl+i from selecting the first input ele in [viewport or page]

and... that's it. for now

thoughts?

**Comment by MikesGlitch** [comment]
Cool. Those changes are made and uploaded. It may take a couple of days for Google to approve and distribute it. Thanks for the suggestion! 

I've made the shortcut Alt+I because we already use the ALT modifier in the other shortcut. You can rebind it though 👍

**Comment by gruns** [comment]
@MikesGlitch ever since our prior discussions here, ive been hyper cognizant on when and where i do want to select the non-first input element. here are two real world examples ive organically come across recently where i said to myself 'huh. mike is right! i *do* want to select the non-first input element! 🙂 

- amazon → i want to focus this input box, not the top input box
![image](https://github.com/user-attachments/assets/b0b3390b-5f3c-4374-a6e5-6e171ec18584)

- google → i want to focus this input box, not the top input box
![image](https://github.com/user-attachments/assets/17101c07-fd36-46f9-966d-1620c86f6f1d)

so while rare(r), this does happen. you were right! 🙂

so lets zoom out and re-analyze what's the 'best' solution here

question for you: what's your typically desired usage ratio?
- 90% of the time you want to focus the first input element in the viewport?
- 10% of the time you want to focus the first input element on the page?

**Comment by gruns** [comment]
@MikesGlitch another example of where i want the non-first input element selected:

![image](https://github.com/user-attachments/assets/b10f5079-bf00-4689-92a2-253b86009cea)

that said, in the above screenshot, selecting the first element in the viewport here would still be the top search box, not the desired google translate `<textarea>`. so if `ctrl+i` iterated through input boxes, two presses of `ctrl+i` would still be required here. which would be acceptable

**Comment by MikesGlitch** [comment]
> that said, in the above screenshot, selecting the first element in the viewport here would still be the top search box, not the desired google translate `<textarea>`. so if `ctrl+i` iterated through input boxes, two presses of `ctrl+i` would still be required here. which would be acceptable

I like the idea of iterating through the inputs. That's one for the todo list. 👍

**Comment by gruns** [comment]
> I like the idea of iterating through the inputs.

👍 

now one question: will the first element selected with the first `ctrl+i` press be the first input element in the viewport? or on the page? if the latter, that input element may be out of the viewport

to me, my most commonly desired behavior remains selection of the first input element on the page. even if its out of the viewport

here's another idea:

- make `ctrl+i` iterate through all the input elements in the viewport, starting with the first in the viewport and iterating down the page. in all the above screenshots, it would take two `ctrl+i` chord presses to arrive at the desired input element, because the first input element on the page, which is *not* the desired input element, also happens to be in the viewport

- add a second bind, `ctrl+/`, which always selects, and scrolls to, the first input element on the page -- whether its currently in the viewport or not

why `ctrl+/`? because `/` is the established shortcut on many sites to select the top/first input element. usually a search box. sites like google, github, and twitter all have `/` (not `ctrl+/`) select the first input element (ie the search box) at the top of the page, even if its out of the viewport. `ctrl+/` is also an unused bind in browsers
